### ISSUE TITLE
Add transport layer for Cloudflare workers

### ISSUE THREAD

**Issue by cupcakearmy** [directus/sdk#29]
_Problem:_ Cloudflare workers don't support native node modules and network requests are made through `fetch` api. Therefore it **does not in cloudflare workers/pages**

In theory the [transport layer is abstracted](https://docs.directus.io/reference/sdk/#custom-implementation-2), but in practise it's not really, as part of the crucial logic is not in the abstracted part (see below) (especially auth related stuff)

https://github.com/directus/sdk/blob/2df91f00971c1d36ea46681829177592ea80a885/src/base/directus.ts#L90-L110

Axios wants to natively support `fetch` in a next release, but god only knows when that will happen xD

Any input? I did not find a "clean" solution to extend the sdk to work without axios. Any ideas? Thanks for the amazing project btw ❤️

**Comment by rijkvanzanten** [comment]
I think this ties back to https://github.com/directus/sdk/issues/26 as well. I kinda wanna drop axios in favor of whatever is natively available (so use `fetch` if it exists, fall back to `XMLHttpRequest` if not. That should save us a lot of headaches, as it effectively turns this module in a dependency free "plain" bit of javascript (to that same extend, I might want to update it from `.ts` files to `.mjs` with TSDoc comments for type checking). Thoughts? 🙂

**Comment by cupcakearmy** [comment]
I'm always for dep-free packages!
Don't know how much further back you want to go with support but `fetch` client side is very much supported `https://caniuse.com/fetch` and is included in node since version 18, so you would only need to polyfill with something like https://github.com/node-fetch/node-fetch

**Comment by cupcakearmy** [comment]
Also to note that this prevents it from being used in service workers, were also only the fetch api is available

**Comment by rijkvanzanten** [comment]
> in node since version 18

I think this is the only problem 🤔 Node 16 is still the LTS, so requiring 18 might be a little premature (although that changes in 2 months, so maybe not?)

**Comment by rijkvanzanten** [comment]
> Also to note that this prevents it from being used in service workers, were also only the fetch api is available

Yeah, we'd have to check if `require` / async `import()` exists in the current globalThis or something like that before attempting to load the module. Long story short, no modules no problems then 😁

**Comment by cupcakearmy** [comment]
Def. version 18 should not be required, but you can lazy load `node-fetch` if the global `fetch` is not available

**Comment by azrikahar** [comment]
Perhaps we can try out https://github.com/unjs/ohmyfetch (used & built by Nuxt team/members). I believe it does the mentioned attempts/checks and try to use native fetch whenever possible 👍

But one little thing is we may lose the uploadProgress provided by Axios (not sure if it's already achievable natively!).

**Comment by rijkvanzanten** [comment]
@azrikahar I think you'd run into the original problem where Cloudflare doesn't allow any node modules to be loaded in the first place, defeating the use case 🙂 

> But one little thing is we may lose the uploadProgress provided by Axios (not sure if it's already achievable natively!).

You can only make network requests in the browser through browser provided APIs, so if axios can do it, so can we 👍🏻

**Comment by rijkvanzanten** [comment]
Linear: [ENG-73](https://linear.app/directus/issue/ENG-73/doesnt-work-in-cloudflare-workers)
